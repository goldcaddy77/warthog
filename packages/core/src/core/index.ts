export * from './BaseModel';
export * from './BaseService';
export * from './config';
export * from './logger';
export * from './types';
