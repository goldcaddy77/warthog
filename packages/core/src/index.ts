export * from './codegen';
export * from './core';
export * from './decorators';
export * from './gql';
export * from './tgql';
export * from './test';
export * from './torm';
export * from './utils';
