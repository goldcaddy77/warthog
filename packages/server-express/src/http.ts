export interface GetResponse {
  statusCode: number;
  body: string;
}

// Custom http `get` without dependencies
export function get(url: string): Promise<GetResponse> {
  // return new pending promise
  return new Promise((resolve, reject) => {
    // select http or https module, depending on reqested url
    const lib = url.startsWith('https') ? require('https') : require('http');
    const request = lib.get(url, (response: any) => {
      // temporary data holder
      const body: string[] = [];
      // on every content chunk, push it to the data array
      response.on('data', (chunk: string) => body.push(chunk));
      // we are done, resolve promise with those joined chunks
      response.on('end', () => {
        return resolve({
          statusCode: response.statusCode,
          body: body.join('')
        });
      });
    });
    // handle connection errors of the request
    request.on('error', (err: Error) => reject(err));
  });
}
