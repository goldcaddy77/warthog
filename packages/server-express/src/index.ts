export * from './context';
export * from './http';
export * from './middleware';
export * from './server';
