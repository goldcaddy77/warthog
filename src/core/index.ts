export * from './server';
export * from './BaseModel';
export * from './config';
export * from './code-generator';
export * from './Context';
export * from './logger';
export * from './types';
export * from './BaseService';
