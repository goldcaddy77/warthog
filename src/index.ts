export * from './core';
export * from './decorators';
export * from './gql';
export * from './middleware';
export * from './schema';
export * from './tgql';
export * from './torm';
export * from './utils';
