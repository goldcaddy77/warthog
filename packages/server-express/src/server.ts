// TODO-MVP: Add custom scalars such as graphql-iso-date
// import { GraphQLDate, GraphQLDateTime, GraphQLTime } from 'graphql-iso-date';

import { ApolloServer, OptionsJson, ApolloServerExpressConfig } from 'apollo-server-express';
import { Request } from 'express';
import express = require('express');
import { GraphQLID, GraphQLSchema } from 'graphql';
import { Binding } from 'graphql-binding';
import { Server as HttpServer } from 'http';
import { Server as HttpsServer } from 'https';
const open = require('open'); // eslint-disable-line @typescript-eslint/no-var-requires
import { AuthChecker, buildSchema } from 'type-graphql'; // formatArgumentValidationError
import { Container } from 'typedi';
import { Connection, ConnectionOptions, useContainer as TypeORMUseContainer } from 'typeorm';

import {
  logger,
  Logger,
  CodeGenerator,
  Config,
  getRemoteBinding,
  createDBConnection
} from '@warthog/core';

import { BaseContext } from './context';
import { DataLoaderMiddleware, healthCheckMiddleware } from './middleware';

import * as Debug from 'debug';

const debug = Debug('warthog:server');

export interface ServerOptions<T> {
  container?: Container;
  apolloConfig?: ApolloServerExpressConfig;
  authChecker?: AuthChecker<T>;
  autoGenerateFiles?: boolean;
  context?: (request: Request) => object;
  expressApp?: express.Application;
  host?: string;
  generatedFolder?: string;
  logger?: Logger;
  middlewares?: any[]; // TODO: fix
  openPlayground?: boolean;
  port?: string | number;
  resolversPath?: string[];
  warthogImportPath?: string;
  introspection?: boolean; // DEPRECATED
  bodyParserConfig?: OptionsJson;
  onBeforeGraphQLMiddleware?: (app: express.Application) => void;
  onAfterGraphQLMiddleware?: (app: express.Application) => void;
}

export class Server<C extends BaseContext> {
  config: Config;
  apolloConfig?: ApolloServerExpressConfig;
  authChecker?: AuthChecker<C>;
  connection!: Connection;
  container: Container;
  expressApp!: express.Application;
  graphQLServer!: ApolloServer;
  httpServer!: HttpServer | HttpsServer;
  logger: Logger;
  schema?: GraphQLSchema;
  bodyParserConfig?: OptionsJson;

  constructor(
    private appOptions: ServerOptions<C>,
    private dbOptions: Partial<ConnectionOptions> = {}
  ) {
    if (typeof this.appOptions.host !== 'undefined') {
      process.env.WARTHOG_APP_HOST = this.appOptions.host;
      // When we move to v2.0 we'll officially deprecate these config values in favor of ENV vars
      // throw new Error(
      //   '`host` option has been removed, please set `WARTHOG_APP_HOST` environment variable instead'
      // );
    }
    if (typeof this.appOptions.port !== 'undefined') {
      process.env.WARTHOG_APP_PORT = this.appOptions.port.toString();
    }
    if (typeof this.appOptions.generatedFolder !== 'undefined') {
      process.env.WARTHOG_GENERATED_FOLDER = this.appOptions.generatedFolder;
    }
    if (typeof this.appOptions.introspection !== 'undefined') {
      process.env.WARTHOG_INTROSPECTION = this.appOptions.introspection ? 'true' : 'false';
    }
    if (typeof this.appOptions.openPlayground !== 'undefined') {
      process.env.WARTHOG_AUTO_OPEN_PLAYGROUND = this.appOptions.openPlayground ? 'true' : 'false';
    }
    if (typeof this.appOptions.autoGenerateFiles !== 'undefined') {
      process.env.WARTHOG_AUTO_GENERATE_FILES = this.appOptions.autoGenerateFiles
        ? 'true'
        : 'false';
    }

    // Ensure that Warthog, TypeORM and TypeGraphQL are all using the same typedi container
    this.container = this.appOptions.container || Container;
    TypeORMUseContainer(this.container as any); // TODO: fix any

    this.authChecker = this.appOptions.authChecker;
    this.bodyParserConfig = this.appOptions.bodyParserConfig;
    this.apolloConfig = this.appOptions.apolloConfig || {};

    this.logger = this.getLogger();

    // NOTE: this should be after we hard-code the WARTHOG_ env vars above because we want the config
    // module to think they were set by the user
    this.config = new Config({ container: this.container, logger: this.logger });

    this.expressApp = this.appOptions.expressApp || express();

    if (!process.env.NODE_ENV) {
      throw new Error("NODE_ENV must be set - use 'development' locally");
    }
  }

  getLogger(): Logger {
    if (this.appOptions.logger) {
      return this.appOptions.logger;
      // } else if (Container.has('warthog.logger')) {
      //   return Container.get('warthog.logger');
    }
    return logger;
  }

  async establishDBConnection(): Promise<Connection> {
    if (!this.connection) {
      debug('establishDBConnection:start');
      this.connection = await createDBConnection(this.dbOptions);
      debug('establishDBConnection:end');
    }

    return this.connection;
  }

  getServerUrl() {
    return `${this.config.get('APP_PROTOCOL')}://${this.config.get('APP_HOST')}:${this.config.get(
      'APP_PORT'
    )}`;
  }

  getGraphQLServerUrl() {
    return `${this.getServerUrl()}/graphql`;
  }

  async getBinding(options: { origin?: string; token?: string } = {}): Promise<Binding> {
    let binding;
    try {
      binding = await getRemoteBinding(this.getGraphQLServerUrl(), {
        origin: 'warthog',
        ...options
      });
      return binding;
    } catch (error) {
      if (error.result && error.result.errors) {
        const messages = error.result.errors.map((item: any) => item.message);
        throw new Error(JSON.stringify(messages));
      }
      throw error;
    }
  }

  async buildGraphQLSchema(): Promise<GraphQLSchema> {
    if (!this.schema) {
      debug('server:buildGraphQLSchema:start');
      this.schema = await buildSchema({
        authChecker: this.authChecker,
        scalarsMap: [
          {
            type: 'ID' as any,
            scalar: GraphQLID
          }
        ],
        container: this.container as any,
        // TODO: ErrorLoggerMiddleware
        globalMiddlewares: [DataLoaderMiddleware, ...(this.appOptions.middlewares || [])],
        resolvers: this.config.get('RESOLVERS_PATH'),
        // TODO: scalarsMap: [{ type: GraphQLDate, scalar: GraphQLDate }]
        validate: this.config.get('VALIDATE_RESOLVERS') === 'true'
      });
      debug('server:buildGraphQLSchema:end');
    }

    return this.schema;
  }

  async generateFiles(): Promise<void> {
    debug('start:generateFiles:start');
    await new CodeGenerator(this.config.get('GENERATED_FOLDER'), this.config.get('DB_ENTITIES'), {
      resolversPath: this.config.get('RESOLVERS_PATH'),
      validateResolvers: this.config.get('VALIDATE_RESOLVERS') === 'true',
      warthogImportPath: this.config.get('MODULE_IMPORT_PATH')
    }).generate();

    debug('start:generateFiles:end');
  }

  private startHttpServer(url: string): void {
    const keepAliveTimeout = Number(this.config.get('WARTHOG_KEEP_ALIVE_TIMEOUT_MS'));
    const headersTimeout = Number(this.config.get('WARTHOG_HEADERS_TIMEOUT_MS'));

    this.httpServer = this.expressApp.listen({ port: this.config.get('APP_PORT') }, () =>
      this.logger.info(`ðŸš€ Server ready at ${url}`)
    );

    this.httpServer.keepAliveTimeout = keepAliveTimeout;
    this.httpServer.headersTimeout = headersTimeout;
  }

  async start() {
    debug('start:start');
    await this.establishDBConnection();
    if (this.config.get('AUTO_GENERATE_FILES') === 'true') {
      await this.generateFiles();
    }
    await this.buildGraphQLSchema();
    const contextGetter =
      this.appOptions.context ||
      (async () => {
        return {};
      });

    debug('start:ApolloServerAllocation:start');
    // See all options here: https://github.com/apollographql/apollo-server/blob/9ffb4a847e1503ea2ab1f3fcd47837daacf40870/packages/apollo-server-core/src/types.ts#L69
    const playgroundOption = this.config.get('PLAYGROUND') === 'true' ? { playground: true } : {};
    const introspectionOption =
      this.config.get('INTROSPECTION') === 'true' ? { introspection: true } : {};

    this.graphQLServer = new ApolloServer({
      context: async (options: { req: Request }) => {
        const consumerCtx = await contextGetter(options.req);

        return {
          connection: this.connection,
          dataLoader: {
            initialized: false,
            loaders: {}
          },
          request: options.req,
          // Allows consumer to add to the context object - ex. context.user
          ...consumerCtx
        };
      },
      ...playgroundOption,
      ...introspectionOption,
      schema: this.schema,
      ...this.apolloConfig
    });

    debug('start:ApolloServerAllocation:end');

    this.expressApp.use('/health', healthCheckMiddleware);

    if (this.appOptions.onBeforeGraphQLMiddleware) {
      this.appOptions.onBeforeGraphQLMiddleware(this.expressApp);
    }

    debug('start:applyMiddleware:start');
    this.graphQLServer.applyMiddleware({
      app: this.expressApp,
      bodyParserConfig: this.bodyParserConfig,
      path: '/graphql'
    });
    debug('start:applyMiddleware:end');

    if (this.appOptions.onAfterGraphQLMiddleware) {
      this.appOptions.onAfterGraphQLMiddleware(this.expressApp);
    }

    const url = this.getGraphQLServerUrl();
    this.startHttpServer(url);

    // Open up websocket connection for subscriptions
    if (this.config.get('SUBSCRIPTIONS') === 'true') {
      this.graphQLServer.installSubscriptionHandlers(this.httpServer);
    }

    // Open playground in the browser
    if (this.config.get('AUTO_OPEN_PLAYGROUND') === 'true') {
      // Assigning to variable and logging to appease linter
      const process = open(url, { wait: false });
      debug('process', process);
    }

    debug('start:end');
    return this;
  }

  async stop() {
    this.logger.info('Stopping HTTP Server');
    this.httpServer.close();
    this.logger.info('Closing DB Connection');
    if (this.connection) await this.connection.close();
  }
}

// Backwards compatability.  This was renamed.
export const App = Server;
