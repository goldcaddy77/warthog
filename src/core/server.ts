// TODO-MVP: Add custom scalars such as graphql-iso-date
// import { GraphQLDate, GraphQLDateTime, GraphQLTime } from 'graphql-iso-date';

import { ApolloServer, OptionsJson, ApolloServerExpressConfig } from 'apollo-server-express';
import { Request } from 'express';
import express = require('express');
import { GraphQLSchema } from 'graphql';
import { Binding } from 'graphql-binding';
import { Server as HttpServer } from 'http';
import { Server as HttpsServer } from 'https';
const open = require('open'); // eslint-disable-line @typescript-eslint/no-var-requires
import { AuthChecker } from 'type-graphql'; // formatArgumentValidationError
import { Container } from 'typedi';
import { Connection, ConnectionOptions, useContainer as TypeORMUseContainer } from 'typeorm';

import { logger, Logger } from '../core/logger';
import { debug } from '../decorators';
import { getRemoteBinding } from '../gql';
import { healthCheckMiddleware } from '../middleware';
import { SchemaBuilder } from '../schema';
import { createDBConnection } from '../torm';

import { CodeGenerator } from './code-generator';
import { Config } from './config';
import { BaseContext } from './Context';

export interface ServerOptions<T> {
  container?: Container;
  apolloConfig?: ApolloServerExpressConfig;
  authChecker?: AuthChecker<T>;
  autoGenerateFiles?: boolean;
  context?: (request: Request) => object;
  expressApp?: express.Application;
  host?: string;
  generatedFolder?: string;
  logger?: Logger;
  middlewares?: any[]; // TODO: fix
  openPlayground?: boolean;
  port?: string | number;
  resolversPath?: string[];
  warthogImportPath?: string;
  introspection?: boolean; // DEPRECATED
  bodyParserConfig?: OptionsJson;
  onBeforeGraphQLMiddleware?: (app: express.Application) => void;
  onAfterGraphQLMiddleware?: (app: express.Application) => void;
}

export class Server<C extends BaseContext> {
  config: Config;
  apolloConfig?: ApolloServerExpressConfig;
  authChecker?: AuthChecker<C>;
  connection!: Connection;
  container: Container;
  expressApp!: express.Application;
  graphQLServer!: ApolloServer;
  httpServer!: HttpServer | HttpsServer;
  logger: Logger;
  schema?: GraphQLSchema;
  bodyParserConfig?: OptionsJson;
  schemaBuilder: SchemaBuilder;

  constructor(
    private appOptions: ServerOptions<C>,
    private dbOptions: Partial<ConnectionOptions> = {}
  ) {
    if (typeof this.appOptions.host !== 'undefined') {
      process.env.WARTHOG_APP_HOST = this.appOptions.host;
      // When we move to v2.0 we'll officially deprecate these config values in favor of ENV vars
      // throw new Error(
      //   '`host` option has been removed, please set `WARTHOG_APP_HOST` environment variable instead'
      // );
    }
    if (typeof this.appOptions.port !== 'undefined') {
      process.env.WARTHOG_APP_PORT = this.appOptions.port.toString();
    }
    if (typeof this.appOptions.generatedFolder !== 'undefined') {
      process.env.WARTHOG_GENERATED_FOLDER = this.appOptions.generatedFolder;
    }
    if (typeof this.appOptions.introspection !== 'undefined') {
      process.env.WARTHOG_INTROSPECTION = this.appOptions.introspection ? 'true' : 'false';
    }
    if (typeof this.appOptions.openPlayground !== 'undefined') {
      process.env.WARTHOG_AUTO_OPEN_PLAYGROUND = this.appOptions.openPlayground ? 'true' : 'false';
    }
    if (typeof this.appOptions.autoGenerateFiles !== 'undefined') {
      process.env.WARTHOG_AUTO_GENERATE_FILES = this.appOptions.autoGenerateFiles
        ? 'true'
        : 'false';
    }

    // Ensure that Warthog, TypeORM and TypeGraphQL are all using the same typedi container
    this.container = this.appOptions.container || Container;
    TypeORMUseContainer(this.container as any); // TODO: fix any

    this.authChecker = this.appOptions.authChecker;
    this.bodyParserConfig = this.appOptions.bodyParserConfig;
    this.apolloConfig = this.appOptions.apolloConfig || {};

    this.logger = this.getLogger();

    // NOTE: this should be after we hard-code the WARTHOG_ env vars above because we want the config
    // module to think they were set by the user
    this.config = Container.get('Config') as Config;
    this.schemaBuilder = Container.get('SchemaBuilder') as SchemaBuilder;

    this.expressApp = this.appOptions.expressApp || express();

    if (!process.env.NODE_ENV) {
      throw new Error("NODE_ENV must be set - use 'development' locally");
    }
  }

  getLogger(): Logger {
    if (this.appOptions.logger) {
      return this.appOptions.logger;
      // } else if (Container.has('warthog.logger')) {
      //   return Container.get('warthog.logger');
    }
    return logger;
  }

  @debug('warthog:server')
  async establishDBConnection(): Promise<Connection> {
    if (!this.connection) {
      this.connection = await createDBConnection(this.dbOptions);
    }

    return this.connection;
  }

  getServerUrl() {
    return `${this.config.get('APP_PROTOCOL')}://${this.config.get('APP_HOST')}:${this.config.get(
      'APP_PORT'
    )}`;
  }

  getGraphQLServerUrl() {
    return `${this.getServerUrl()}/graphql`;
  }

  async getBinding(options: { origin?: string; token?: string } = {}): Promise<Binding> {
    let binding;
    try {
      binding = await getRemoteBinding(this.getGraphQLServerUrl(), {
        origin: 'warthog',
        ...options
      });
      return binding;
    } catch (error) {
      if (error.result && error.result.errors) {
        const messages = error.result.errors.map((item: any) => item.message);
        throw new Error(JSON.stringify(messages));
      }
      throw error;
    }
  }

  @debug('warthog:server')
  async buildGraphQLSchema(): Promise<GraphQLSchema> {
    if (!this.schema) {
      this.schema = await this.schemaBuilder.build({
        authChecker: this.authChecker,
        middlewares: this.appOptions.middlewares
      });
    }

    return this.schema;
  }

  // @debug
  async generateFiles(): Promise<void> {
    const generator = Container.get('CodeGenerator') as CodeGenerator;
    return generator.generate();
  }

  private startHttpServer(url: string): void {
    const keepAliveTimeout = Number(this.config.get('WARTHOG_KEEP_ALIVE_TIMEOUT_MS'));
    const headersTimeout = Number(this.config.get('WARTHOG_HEADERS_TIMEOUT_MS'));

    this.httpServer = this.expressApp.listen({ port: this.config.get('APP_PORT') }, () =>
      this.logger.info(`ðŸš€ Server ready at ${url}`)
    );

    this.httpServer.keepAliveTimeout = keepAliveTimeout;
    this.httpServer.headersTimeout = headersTimeout;
  }

  @debug('warthog:server')
  async start() {
    await this.establishDBConnection();
    if (this.config.get('AUTO_GENERATE_FILES') === 'true') {
      await this.generateFiles();
    }
    await this.buildGraphQLSchema();

    const contextGetter =
      this.appOptions.context ||
      (async () => {
        return {};
      });

    // See all options here: https://github.com/apollographql/apollo-server/blob/9ffb4a847e1503ea2ab1f3fcd47837daacf40870/packages/apollo-server-core/src/types.ts#L69
    const playgroundOption = this.config.get('PLAYGROUND') === 'true' ? { playground: true } : {};
    const introspectionOption =
      this.config.get('INTROSPECTION') === 'true' ? { introspection: true } : {};

    this.graphQLServer = new ApolloServer({
      context: async (options: { req: Request }) => {
        const consumerCtx = await contextGetter(options.req);

        return {
          connection: this.connection,
          dataLoader: {
            initialized: false,
            loaders: {}
          },
          request: options.req,
          // Allows consumer to add to the context object - ex. context.user
          ...consumerCtx
        };
      },
      ...playgroundOption,
      ...introspectionOption,
      schema: this.schema,
      ...this.apolloConfig
    });

    this.expressApp.use('/health', healthCheckMiddleware);

    if (this.appOptions.onBeforeGraphQLMiddleware) {
      this.appOptions.onBeforeGraphQLMiddleware(this.expressApp);
    }

    this.graphQLServer.applyMiddleware({
      app: this.expressApp,
      bodyParserConfig: this.bodyParserConfig,
      path: '/graphql'
    });

    if (this.appOptions.onAfterGraphQLMiddleware) {
      this.appOptions.onAfterGraphQLMiddleware(this.expressApp);
    }

    const url = this.getGraphQLServerUrl();
    this.startHttpServer(url);

    // Open up websocket connection for subscriptions
    if (this.config.get('SUBSCRIPTIONS') === 'true') {
      this.graphQLServer.installSubscriptionHandlers(this.httpServer);
    }

    // Open playground in the browser
    if (this.config.get('AUTO_OPEN_PLAYGROUND') === 'true') {
      // Assigning to variable and logging to appease linter
      open(url, { wait: false });
    }

    return this;
  }

  async stop() {
    this.logger.info('Stopping HTTP Server');
    this.httpServer.close();
    this.logger.info('Closing DB Connection');
    await this.connection.close();
  }
}

// Backwards compatability.  This was renamed.
export const App = Server;
