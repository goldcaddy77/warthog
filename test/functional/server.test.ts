/* eslint-disable @typescript-eslint/camelcase */

import * as fs from 'fs';
import * as path from 'path';

import { get, GetResponse } from '../../src/core/http';
import { Server } from '../../src/core/server';
import { getBindingError, logger } from '../../src';

// import { Binding } from '../../test/generated/binding';
import { getTestServer } from '../test-server';
import { KitchenSink } from '../modules';
import { KITCHEN_SINKS } from './fixtures';

import { setTestServerEnvironmentVariables } from '../server-vars';

let server: Server<any>;
// let binding: Binding;
let binding: any;

describe('server', () => {
  beforeEach(() => {
    jest.setTimeout(20000);
  });

  // Make sure to clean up server
  beforeAll(async done => {
    jest.setTimeout(20000);
    await cleanUpTestData();

    try {
      setTestServerEnvironmentVariables();

      server = getTestServer({
        apolloConfig: { playground: false }
      });

      await server.start();

      binding = (await server.getBinding()) as unknown;
    } catch (error) {
      logger.error(error);
      process.exit(1);
    }

    const kitchenSink = await createKitchenSink(binding, 'hi@warthog.com');
    await createManyDishes(binding, kitchenSink.id);
    await createManyKitchenSinks(binding);

    done();
  });

  // Make sure to clean up server
  afterAll(async done => {
    await cleanUpTestData();
    await server.stop();
    done();
  });

  test('disables playground properly using apollo config options', async () => {
    expect.assertions(2);
    const response: GetResponse = await get(server.getGraphQLServerUrl());

    expect(response.statusCode).toEqual(400);
    expect(response.body).toContain('GET query missing');
  });

  test('queries deeply nested objects', async () => {
    expect.assertions(2);
    const results = await binding.query.kitchenSinks(
      { skip: 0, orderBy: 'createdAt_ASC', limit: 1 },
      `{
          stringField
          emailField
          integerField
          booleanField
          floatField
          dishes {
            name
            kitchenSink {
              stringField
            }
          }
          createdById
          updatedById
          version
        }`
    );

    expect(results).toMatchSnapshot();
    expect(results[0].dishes.length).toEqual(20);
  });

  test('throws errors when given bad input', async done => {
    expect.assertions(1);

    createKitchenSink(binding, '')
      .catch(error => {
        expect(error).toBeDefined();
      })
      .finally(done);
  });

  test('getBindingError pulls correct info from binding error', async done => {
    expect.assertions(4);

    let originalError;
    let improvedError;
    try {
      await createKitchenSink(binding, 'not.an.email');
    } catch (err) {
      originalError = err;
      improvedError = getBindingError(err);
    }

    // Ensure that the raw error is useless. If this ever breaks and gives back a lot of
    // useful info, we can get rid of getBindingError and just use the vanilla error
    expect(Object.keys(originalError)).toEqual(['message', 'locations', 'path']);

    // Note: this should likely not be an INTERNAL_SERVER_ERROR since it's based on bad user input
    expect(improvedError.message).toEqual('Argument Validation Error');

    // TODO: this should likely be cleaned up
    expect(improvedError.validationErrors.emailField.isEmail).toEqual(
      'emailField must be an email'
    );

    expect(improvedError.validationErrors).toMatchSnapshot();

    done();
  });

  test('string query: exact match (Nakia)', async () => {
    expect.assertions(2);

    const result = await binding.query.kitchenSinks(
      { where: { stringField_eq: 'Nakia' } },
      '{ stringField }'
    );
    expect(result.length).toEqual(1);
    expect(result).toMatchSnapshot();
  });

  test('string query: contains `a` (upper or lower)', async () => {
    expect.assertions(2);

    const result = await binding.query.kitchenSinks(
      { where: { stringField_contains: 'a' } },
      '{ stringField }'
    );
    expect(result.length).toEqual(50);
    expect(result).toMatchSnapshot();
  });

  test('string query: starts with `b` (upper or lower)', async () => {
    expect.assertions(2);

    const result = await binding.query.kitchenSinks(
      { where: { stringField_startsWith: 'b' } },
      '{ stringField }'
    );
    expect(result.length).toEqual(5);
    expect(result).toMatchSnapshot();
  });

  test('string query: ends with `z` (upper or lower)', async () => {
    expect.assertions(2);

    const result = await binding.query.kitchenSinks(
      { where: { stringField_endsWith: 'z' } },
      '{ stringField }'
    );
    expect(result.length).toEqual(3);
    expect(result).toMatchSnapshot();
  });

  test('string query: in list { devin, erling, KAELYN, raquel }', async () => {
    expect.assertions(2);

    const result = await binding.query.kitchenSinks(
      { where: { stringField_in: ['devin', 'erling', 'KAELYN', 'raquel'] } },
      '{ stringField }'
    );
    expect(result.length).toEqual(4);
    expect(result).toMatchSnapshot();
  });
});

async function createKitchenSink(binding: any, email: string): Promise<KitchenSink> {
  return binding.mutation.createKitchenSink(
    {
      data: {
        stringField: 'My String',
        emailField: email,
        integerField: 123,
        booleanField: true,
        floatField: 123.456
      }
    },
    `{ id }`
  );
}

async function createManyKitchenSinks(binding: any): Promise<KitchenSink> {
  const data = KITCHEN_SINKS.map(item => {
    const { dateField, ...fields } = item; // eslint-disable-line
    return fields;
  });

  return binding.mutation.createManyKitchenSinks({ data }, `{ id }`);
}

async function createManyDishes(binding: any, kitchenSinkId: string): Promise<KitchenSink> {
  const data = Array.from({ length: 20 }, (v, i) => i).map(item => {
    return {
      name: `Dish ${item}`,
      kitchenSinkId
    };
  });
  let dishes;

  try {
    dishes = await binding.mutation.createManyDishs({ data }, `{ id name createdById }`);
  } catch (error) {
    logger.error(error);
  }

  return dishes;
}

async function cleanUpTestData() {
  try {
    fs.unlinkSync(path.join(__dirname, '../../warthog.sqlite.tmp'));
  } catch (error) {
    // console.error('Error cleaning up test data', error);
  }

  return;
}

/* eslint-enable @typescript-eslint/camelcase */
