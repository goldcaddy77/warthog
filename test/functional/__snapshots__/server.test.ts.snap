// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server getBindingError pulls correct info from binding error 1`] = `
Object {
  "emailField": Object {
    "isEmail": "emailField must be an email",
  },
}
`;

exports[`server queries deeply nested objects 1`] = `
Array [
  Object {
    "booleanField": true,
    "createdById": "abc123",
    "dishes": Array [
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 0",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 1",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 2",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 3",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 4",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 5",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 6",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 7",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 8",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 9",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 10",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 11",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 12",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 13",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 14",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 15",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 16",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 17",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 18",
      },
      Object {
        "kitchenSink": Object {
          "stringField": "My String",
        },
        "name": "Dish 19",
      },
    ],
    "emailField": "hi@warthog.com",
    "floatField": 123.456,
    "integerField": 123,
    "stringField": "My String",
    "updatedById": null,
    "version": 1,
  },
]
`;

exports[`server string query: contains \`a\` (upper or lower) 1`] = `
Array [
  Object {
    "stringField": "Trantow",
  },
  Object {
    "stringField": "VOLUPTATUM NULLA",
  },
  Object {
    "stringField": "maximillia",
  },
  Object {
    "stringField": "QUAM CULPA",
  },
  Object {
    "stringField": "Madeline",
  },
  Object {
    "stringField": "NAKIA",
  },
  Object {
    "stringField": "nakia",
  },
  Object {
    "stringField": "Nakia",
  },
  Object {
    "stringField": "sed praesentium",
  },
  Object {
    "stringField": "amber",
  },
  Object {
    "stringField": "delfinaz",
  },
  Object {
    "stringField": "SHEMAR",
  },
  Object {
    "stringField": "stark",
  },
  Object {
    "stringField": "elda",
  },
  Object {
    "stringField": "Damien",
  },
  Object {
    "stringField": "earum nesciunt",
  },
  Object {
    "stringField": "illum atque",
  },
  Object {
    "stringField": "schowalter",
  },
  Object {
    "stringField": "MAIORES ADIPISCI",
  },
  Object {
    "stringField": "VELIT ADIPISCI",
  },
  Object {
    "stringField": "AMET QUI",
  },
  Object {
    "stringField": "maiores praesentium",
  },
  Object {
    "stringField": "KAELYN",
  },
  Object {
    "stringField": "alias sint",
  },
  Object {
    "stringField": "molestiae praesentium",
  },
  Object {
    "stringField": "Okuneva",
  },
  Object {
    "stringField": "hartmann",
  },
  Object {
    "stringField": "raquel",
  },
  Object {
    "stringField": "quas fugit",
  },
  Object {
    "stringField": "quia et",
  },
  Object {
    "stringField": "ut consequatur",
  },
  Object {
    "stringField": "iusto perspiciatis",
  },
  Object {
    "stringField": "HAMILL",
  },
  Object {
    "stringField": "LABORE CULPA",
  },
  Object {
    "stringField": "VERITATIS EIUS",
  },
  Object {
    "stringField": "HELGA",
  },
  Object {
    "stringField": "ipsum voluptas",
  },
  Object {
    "stringField": "linnea",
  },
  Object {
    "stringField": "madison",
  },
  Object {
    "stringField": "NUMQUAM ALIQUAM",
  },
  Object {
    "stringField": "macy",
  },
  Object {
    "stringField": "GERLACH",
  },
  Object {
    "stringField": "padberg",
  },
  Object {
    "stringField": "ENIM DICTA",
  },
  Object {
    "stringField": "amalia",
  },
  Object {
    "stringField": "QUAS PERSPICIATIS",
  },
  Object {
    "stringField": "Williamson",
  },
  Object {
    "stringField": "schamberger",
  },
  Object {
    "stringField": "Jared",
  },
  Object {
    "stringField": "QUIDEM FUGIAT",
  },
]
`;

exports[`server string query: ends with \`z\` (upper or lower) 1`] = `
Array [
  Object {
    "stringField": "LUEILWITZ",
  },
  Object {
    "stringField": "delfinaz",
  },
  Object {
    "stringField": "horowitz",
  },
]
`;

exports[`server string query: exact match (Nakia) 1`] = `
Array [
  Object {
    "stringField": "Nakia",
  },
]
`;

exports[`server string query: in list { devin, erling, KAELYN, raquel } 1`] = `
Array [
  Object {
    "stringField": "devin",
  },
  Object {
    "stringField": "erling",
  },
  Object {
    "stringField": "KAELYN",
  },
  Object {
    "stringField": "raquel",
  },
]
`;

exports[`server string query: starts with \`b\` (upper or lower) 1`] = `
Array [
  Object {
    "stringField": "BOSCO",
  },
  Object {
    "stringField": "BERENICE",
  },
  Object {
    "stringField": "billie",
  },
  Object {
    "stringField": "Block",
  },
  Object {
    "stringField": "BAUCH",
  },
]
`;
